File: lab06a.py
#!/usr/bin/env python3
"""
Programming Assignment 6: Quicksort
"""


import random


def quicksort(n, i, j):
    """
    Recursively sort a list of numbers using the Quicksort algorithm
    """
    left = i
    right = j
    pick = n[i]
    while left <= right:
        while n[left] < pick:  # adjust left until an element is out of place
            left += 1
        while n[right] > pick:  # adjust right until an element is out of place
            right -= 1
        if left <= right:  # interchange the elements
            (n[left], n[right]) = (n[right], n[left])
            left += 1
            right -= 1
    if right > i:  # recurse on smaller problems
        quicksort(n, i, right)
    if left < j:
        quicksort(n, left, j)


def test():
    """
    Test cases.
    """
    for size in range(1, 30):
        numbers = []
        for _ in range(size):
            numbers.append(random.randrange(size))
        expected = sorted(numbers)
        result = numbers.copy()
        quicksort(result, 0, len(result) - 1)
        if result != expected:
            print('   input', numbers)
            print('  result', result)
            print('expected', expected)
    numbers = []
    print('Enter the elements of your array, all on one line')
    for word in input().split():
        numbers.append(int(word))
    quicksort(numbers, 0, len(numbers) - 1)
    print('Sorted numbers are ', numbers)


test()

File: lab06a.cpp
//* Translate the PASCAL program described in the file lab06a.md into C++
//* then find and fix the bugs

//* Fix the bugs in the file lab06a.cpp

#include <iostream>
#include <vector>

using namespace std;

const int MAX_SIZE = 15;

void swap_elements(vector<int>& n, int& left, int& right) {
    int temp = n[left];
    n[left] = n[right];
    n[right] = temp;
    left++;
    right--;
}

void quicksort(vector<int>& n, int low, int high) {
    if (low < high) {
        // Lomuto partition: choose pivot as n[high]
        int pivot = n[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (n[j] <= pivot) {
                i++;
                // Simple swap without modifying indices
                int temp = n[i];
                n[i] = n[j];
                n[j] = temp;
            }
        }
        // Swap pivot to correct position
        int temp = n[i + 1];
        n[i + 1] = n[high];
        n[high] = temp;
        int pi = i + 1;
        quicksort(n, low, pi - 1);
        quicksort(n, pi + 1, high);
    }
}

int main() {
    cout << "I can sort arrays of size up to " << MAX_SIZE << endl;
    cout << "how large an array would you like me to sort? " << endl;
    int arraysize;
    cin >> arraysize;
    if (arraysize > MAX_SIZE || arraysize < 1) {
        cout << "Invalid size" << endl;
        return 1;
    }
    vector<int> n(arraysize);
    cout << "now enter the elements of your array, all on one line " << endl;
    for (int k = 0; k < arraysize; k++) {
        cin >> n[k];
    }
    quicksort(n, 0, arraysize - 1);
    cout << "sorted numbers are " << endl;
    for (int k = 0; k < arraysize; k++) {
        cout << n[k] << " ";
    }
    cout << endl;
    return 0;
}
